### ISSUE TITLE
getting different results from verify and find function.

### ISSUE THREAD

**Issue by Raghucharan16** [serengil/deepface#1196]
I'm trying to extract unique faces from set of folders and I initially saved all the faces in their respective folders and trying to update another folder say unique_faces, Now the situation arises if the newly coming image is in the unique_faces folder or not, so I tried find function and verify function [nested loop  for verify of course] and i verify function being taking time, gives me desired results but not quicker and find method returns the results very fast but not accurate like some of the faces are also getting missed. I don't have issue with duplicates and I tried with all models and by adjusting the threshold distance. But verify and find should return same results on same threshold right?? that too is not working. even the find function calculates the newly added images representations, it is a lot quicker because may be one to three images will be added.  I have the code too.
My only concern is how can improve find method's  accuracy maintaining the fastness.

find method's code,
```
def verify_faces_in_database_2(face_folder, database_folder, hashmap):
    logging.info('Starting verification in database in folder %s', face_folder)
    db_filename = None
    face_image_path = None
    try:
        face_files = os.listdir(face_folder)
        for face_filename in face_files:
            if face_filename.startswith('.'):
                continue  # Skip system files like .DS_Store
            face_image_path = os.path.join(face_folder, face_filename)
            # for db_filename in os.listdir(database_folder):
            #     if db_filename.startswith('.'):
            #         continue  # Skip system files like .DS_Store
                #db_image_path = os.path.join(database_folder, db_filename)
                #if os.path.isfile(db_image_path):
                    # logging.info('face image path: %s', face_image_path)
                    #print('face image path: %s', face_image_path)
            verified = DeepFace.find(img1_path=face_image_path, db_path=database_folder,
                                        model_name='VGG-Face',
                                        enforce_detection=False)
            
            # logging.info('verified: %s', verified)
            distance_threshold = 0.5
            if verified[0]['distance'] < distance_threshold:
                logging.info(f"Match found between {db_filename} and folder name {face_folder} and {face_filename}")
                logging.info(verified[0])
                break  # Stop searching for matches if one is found
            # else:
            #     verified['verified'] = False

            # if verified['verified']:
            #     # print('verified: ', verified)
                
            else:
                logging.info('No match found, saving the image in the database with a unique ID')
                unique_id = str(uuid.uuid4())  # Generate unique ID
                # Save the image in the database with the unique ID
                save_to_database_with_unique_id(face_image_path, unique_id, database_folder)
                call_aws_api(face_image_path, hashmap, database_folder)
```

verify's code

```
def verify_faces_in_database_2(face_folder, database_folder, hashmap):
    logging.info('Starting verification in database in folder %s', face_folder)
    db_filename = None
    face_image_path = None
    try:
        face_files = os.listdir(face_folder)
        for face_filename in face_files:
            if face_filename.startswith('.'):
                continue  # Skip system files like .DS_Store
            face_image_path = os.path.join(face_folder, face_filename)
            for db_filename in os.listdir(database_folder):
                if db_filename.startswith('.'):
                    continue  # Skip system files like .DS_Store
                db_image_path = os.path.join(database_folder, db_filename)
                if os.path.isfile(db_image_path):
                    # logging.info('face image path: %s', face_image_path)
                    print('face image path: %s', face_image_path)
                    verified = DeepFace.verify(img1_path=db_image_path, img2_path=face_image_path,
                                               model_name='Dlib',
                                               enforce_detection=False)
                    # logging.info('verified: %s', verified)
                    distance_threshold = 0.50
                    if verified['distance'] < distance_threshold:
                        verified['verified'] = True
                    else:
                        verified['verified'] = False

                    if verified['verified']:
                        # print('verified: ', verified)
                        logging.info(
                            f"Match found between {db_filename} and folder name {face_folder} and {face_filename}")
                        break  # Stop searching for matches if one is found
            else:
                logging.info('No match found, saving the image in the database with a unique ID')
                unique_id = str(uuid.uuid4())  # Generate unique ID
                # Save the image in the database with the unique ID
                save_to_database_with_unique_id(face_image_path, unique_id, database_folder)
                call_aws_api(face_image_path, hashmap, database_folder)
```

**Comment by serengil** [comment]
you are using different models in verify and find functions. one is vgg, other is dlib.

TLDR: find and verify do the same, just find stores already calculated thresholds in a pickle file to find faster in next run.

**Comment by Raghucharan16** [comment]
My bad I pasted wrong code, I tried with the same models and I edited the same, with same model, with the same threshold results are different. i think this needs a  fix.

**Comment by serengil** [comment]
if a photo has many faces, then verify considers the most similar one between the other photo.

on the other hand, find calculates distances for each face. this is the only difference.

**Comment by Raghucharan16** [comment]
ok. but i'm extracting all the faces in an image and saving them as individual images into a folder. and using this folder i'm checking the unique_faces folder. so there aren't any mutliple faces in a single image.

**Comment by serengil** [comment]
you may try to print distance values between image pairs. if you see any difference on these distance values for same image pair, we can continue to discuss.

**Comment by Raghucharan16** [comment]
yes...that is what I'm exactly trying to say...
here is the code for verify function
```
import os
from deepface import DeepFace

def verify_image_against_folder(image_path, folder_path):
    try:
        verified_results = {}
        for filename in os.listdir(folder_path):
            if filename.endswith(('.jpg', '.jpeg', '.png')):
                image_to_compare_path = os.path.join(folder_path, filename)
                result = DeepFace.verify(image_path, image_to_compare_path)
                if result["verified"]:
                    verified_results[filename] = result
    except:
        pass
    return verified_results

# Usage example:
reference_image_path = "Madhursample.jpg"
folder_path_to_compare = "dbmod"

results = verify_image_against_folder(reference_image_path, folder_path_to_compare)

# Print results
for filename, result in results.items():
    print(f"Verification result for {filename}: {result['verified']}")
    print(f"Distance: {result['distance']}")
```
and the output is 
```
Verification result for AdityaVikramBudholiya_EMqaTest-3$00-0000-0000-000000024758_face.jpg: True
Distance: 0.6586338087844334
Verification result for VivekKumarGupta_EMqaTest-3$00-0000-0000-000000024832_face.jpg: True
Distance: 0.6253741569539091
```

and for find function the code is 
```
from deepface import DeepFace
result=DeepFace.find("Madhursample.jpg",db_path="dbmod",enforce_detection=False)
print(result)
```
and the output is 
```
[                                            identity                                      hash  target_x  target_y  ...  source_w  source_h  threshold  distance
0  dbmod/AlokKumar_EMqaTest-3$00-0000-0000-000000...  de05dfb9108ac362273ee2f033e1db9a961cd360        47       209  ...        73        73       0.68  0.533479
1  dbmod/VivekKumarGupta_EMqaTest-3$00-0000-0000-...  69e1254b38d6875b382ffb4012261f98761fcc06        50        60  ...        73        73       0.68  0.625374
2  dbmod/ManishRanjan_EMqaTest-3$00-0000-0000-000...  3c107ea80669c3a2c37526ef9a102e505d398905        88       276  ...        73        73       0.68  0.639899
3                              dbmod/Madhur_face.jpg  aaa8fe54fc30574dba2ce19ec28eb60fa49fc49b        47        50  ...        73        73       0.68  0.640093
4  dbmod/AdityaVikramBudholiya_EMqaTest-3$00-0000...  ff05c12f927aa766d5834d9015e01af5cee31be1        32        77  ...        73        73       0.68  0.658634
5  dbmod/SiddarthJain_EMqaTest-3$00-0000-0000-000...  32eab590a3ddec2225fa854a2a4e6596fcc644ac        37       176  ...        73        73       0.68  0.679664

[6 rows x 12 columns]]
```
As you can see the clear difference between the 2 results even though input image and folder are same, model [by default is VGG-face] so only difference is methods.

**Comment by serengil** [comment]
would you please share the image pair here?

**Comment by Raghucharan16** [comment]
sure...
this is Madhursample.jpg
![Madhur_face](https://github.com/serengil/deepface/assets/104614903/57646c02-63a0-4967-b618-c2f55ffea4b7)

and for
AdityaVikramBudholiya_EMqaTest-3$00-0000-0000-000000024758_face.jpg
![AdityaVikramBudholiya_EMqaTest-3$00-0000-0000-000000024758_face](https://github.com/serengil/deepface/assets/104614903/a4601124-4dec-47d6-9f31-bfcf880c2f9d)

and AlokKumar_EMqaTest-3$00-0000-0000-0000002468_face.jpg is
![ArunKumarS_EMqaTest-3$00-0000-0000-000000024778_face](https://github.com/serengil/deepface/assets/104614903/30350921-baec-4c41-a9e6-c28b8dd7553e)

**Comment by serengil** [comment]
As I see from your results, find and verify are returning same. What is different?

`Madhursample.jpg` vs `AdityaVikramBudholiya_EMqaTest-3$00-0000-0000-000000024758_face.jpg`
Verify: 0.6586338087844334
Find: 0.658634

`Madhursample.jpg` vs `VivekKumarGupta_EMqaTest-3$00-0000-0000-000000024832_face.jpg`
Verify: 0.6253741569539091
Find: 0.625374

**Comment by Raghucharan16** [comment]
distances are same but the results are not. That is my concern..

**Comment by serengil** [comment]
if distances are same, then there is nothing wrong from the library's perspective.

the procedure to list images in your implementation and find function are different. for instance, if an image is webp but its extension is jpg/png, then this is discarded in find function but it is not discarded in your implementation.

https://github.com/serengil/deepface/blob/master/deepface/commons/image_utils.py#L16

**Comment by Raghucharan16** [comment]
No, I took care of that already, I have are only .jpg images.

**Comment by Raghucharan16** [comment]
May be using insightface would give me better results as it  got better accuracy on south asian and east asian faces..

**Comment by serengil** [comment]
definetely i will. we will add it because of its accuracy for asian faces.

btw - feel free to continue discussion this thread. the reason I closed it that I only set tickets open only if that requires code change.

**Comment by serengil** [comment]
Most probably, your logic is buggy because verify gives same distances with find when tested directly.

**Comment by serengil** [comment]
For instance, listdir does not list files in subfolders. But find function considers sub folders. You are running a different logic.
